### PR TITLE
Refactor Spark job storage configuration panels

### PR DESCRIPTION
**Pull request by wezhang**

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/ui/util/SwingExts.kt [hunk 1]
```diff
@@ -24,15 +24,17 @@ package com.microsoft.intellij.ui.util
 
 import javax.swing.ListModel
 
-val <T> ListModel<T>.iterator: Iterator<T>
-    get() = this.let {
+val <T> ListModel<T>?.iterator: Iterator<T>
+    get() = if (this != null) {
```
Comment (t-rufang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/ui/util/SwingExts.kt]: How about writing in the following way?
```
val <T> ListModel<T>?.iterator: Iterator<T>
    get() = this?.let {
        object: Iterator<T> {
            private var position: Int = 0

            override fun hasNext() = size > position
            override fun next() = getElementAt(position++)
        }
    } ?: emptyList<T>().iterator()

```

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionJobUploadStoragePanel.kt [hunk 2]
```diff
@@ -174,24 +140,97 @@ open class SparkSubmissionJobUploadStoragePanel
...
+                    .doOnNext {
+                        runInEdt(stateForComponent(view)) {
+                            log().info("Current model storage account type is $deployStorageTypeSelection")
+                            updateStorageTypesModelForCluster(it)
+                            log().info("Update model storage account type to $deployStorageTypeSelection")
+
+                            currentCardViewModel?.cluster = it
+                        }
+                    }
+                    .subscribe()
```
Comment (t-rufang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionJobUploadStoragePanel.kt]: Better to handle errors when we `subscribe` a `Subject`

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionJobUploadStorageGen2Card.kt [hunk 3]
```diff
@@ -75,18 +88,59 @@ class SparkSubmissionJobUploadStorageGen2Card : SparkSubmissionJobUploadStorageB
...
+                val rootPathText = gen2RootPathField.text?.trim()
+                return if (AbfsUri.isType(rootPathText)) AbfsUri.parse(rootPathText) else null
+            }
+
+        override var cluster: IClusterDetail?
+            get() = super.cluster
+            set(cluster) {
+                val gen2Account = cluster?.storageAccount?.name ?: return
+
+                if (!storageKeyField.text.isNullOrBlank()) {
```
Comment (t-rufang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionJobUploadStorageGen2Card.kt]: Why do we need to set storage key field when we set cluster? I suppose this behaviour should be done in `setData` method.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/ui/util/SwingExts.kt (hunk 1 above)
Comment (wezhang) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/ui/util/SwingExts.kt]: I've tried this before, there must be a compiler or class loader issue here:
```
java.lang.NoClassDefFoundError: com/microsoft/intellij/ui/util/SwingExtsKt$iterator$1$1

	at com.microsoft.intellij.ui.util.SwingExtsKt.getIterator(SwingExts.kt:28)
	at com.microsoft.intellij.ui.util.SwingExtsKt.findFirst(SwingExts.kt:37)
```

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionJobUploadStoragePanel.kt (hunk 2 above)
Comment (wezhang) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionJobUploadStoragePanel.kt]: Good catch! Thanks!

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionJobUploadStorageGen2Card.kt (hunk 3 above)
Comment (wezhang) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionJobUploadStorageGen2Card.kt]: Here the setData/getData have a role changes in MVVM. They are used for exchanging between Model and ViewModel ( Model <=> ViewModel). The ViewModel can operate view directly.

**Pull request merged**